pandas/_libs/src/inference.pyx

```diff
@@ -606,9 +611,16 @@ cpdef bint is_bool_array(ndarray values):
         if n == 0:
             return False
 
-        for i in range(n):
-            if not util.is_bool_object(objbuf[i]):
-                return False
+        if skipna:
+            for i in range(n):
+                val = objbuf[i]
+                if not util._checknull(val) and not util.is_bool_object(val):
+                    return False
```

Comment:
Refactoring the entire set of functions is kind of a can of worms, because the fundamental problem here is that we must assume that arrays contain potentially mixed types or objects, which are really the same thing.

It might be possible to make a `Validator` class with subclasses for each type that we want to infer, that has a method `validate` that encapsulates the current loop logic.

Generally agree this needs some TLC.